File: Aug/Day21/count-submatrices-with-all-ones.cc
class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        int res = 0;
        vector<vector<int>> dp(m, vector<int>(n, 0));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (j == 0) {
                    dp[i][j] = mat[i][j];
                } else {
                    dp[i][j] = (mat[i][j] == 0) ? 0 : dp[i][j - 1] + 1;
                }

                int cur = dp[i][j];
                for (int k = i; k >= 0; k--) {
                    cur = min(cur, dp[k][j]);
                    if (cur == 0) break;
                    res += cur;
                }
            }
        }
        return res;
    }
};


class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        vector<int> heights(n, 0);
        int res = 0;

        for (auto& row : mat) {
            for (int i = 0; i < n; i++) {
                heights[i] = (row[i] == 0) ? 0 : heights[i] + 1;
            }

            vector<array<int,3>> stack = {{-1, 0, -1}}; // {index, curSum, height}
            for (int i = 0; i < n; i++) {
                int h = heights[i];
                while (!stack.empty() && stack.back()[2] >= h) {
                    stack.pop_back();
                }
                auto [j, prev, _] = stack.back();
                int cur = prev + (i - j) * h;
                stack.push_back({i, cur, h});
                res += cur;
            }
        }
        return res;
    }
};

File: Aug/Day21/count-submatrices-with-all-ones.py
class Solution:
    #TC- O(m2*n), SC - O(m*n)
    def numSubmat(self, mat: List[List[int]]) -> int:
        m, n = len(mat), len(mat[0])
        res = 0
        dp = [[0] * n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if j == 0:
                    dp[i][j] = mat[i][j]
                else:
                    dp[i][j] = 0 if mat[i][j] == 0 else dp[i][j - 1] + 1
                cur = dp[i][j]
                for k in range(i, -1, -1):
                    cur = min(cur, dp[k][j])
                    if cur == 0:
                        break
                    res += cur
        return res

class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        heights = [0] * len(mat[0])
        res = 0
        for row in mat:
            for i, x in enumerate(row):
                heights[i] = 0 if x == 0 else heights[i] + 1
            stack = [[-1, 0, -1]]
            for i, h in enumerate(heights):             
                while stack[-1][2] >= h:
                    stack.pop()
                j, prev, _ = stack[-1]
                cur = prev + (i - j) * h
                stack.append([i, cur, h])
                res += cur
        return res